File: app/src/main/java/kr/blogspot/charlie0301/wimple/FinancialStateSummaryFragment.java
package kr.blogspot.charlie0301.wimple;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Message;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.mikephil.charting.charts.PieChart;

import java.util.Collection;

import kr.blogspot.charlie0301.wimple.WimpleActivity.CommandID;
import kr.blogspot.charlie0301.wimple.impl.WimpleImpl;
import kr.blogspot.charlie0301.wimple.impl.util.ChartUtils;
import kr.blogspot.charlie0301.wimple.impl.util.DateFormatUtils;
import kr.blogspot.charlie0301.wimple.model.AccountState;

public class FinancialStateSummaryFragment  extends Fragment implements IWimpleFragment{

	//private final static String LOG_TAG = "TransactionInsertFragment";

	private final WimpleImpl wimple = WimpleImpl.getInstance();
	private View view = null;
	private Context context = null;

	// GUI
	private LinearLayout llChart;
	
	private LinearLayout llUpdateNotice;
	private TextView tvSavingValue;
	private TextView tvDebtValue;
	private TextView tvSumValue;

	// Data
	private boolean firstUpdate;
	
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
			Bundle savedInstanceState) {

		context = WimpleActivity.context.get();

		view = inflater.inflate(R.layout.fragment_finalcial_state_summary_tab, container, false);
		
		tvSumValue = (TextView)view.findViewById(R.id.as_sum_value);
		tvSavingValue = (TextView)view.findViewById(R.id.as_saving_value);
		tvDebtValue = (TextView)view.findViewById(R.id.as_debt_value);

		llUpdateNotice = (LinearLayout)view.findViewById(R.id.as_update_notification);
		
		firstUpdate = true;
		wimple.getFinancialState(DateFormatUtils.getServerDateString(""), false);
		
		view.findViewById(R.id.as_refresh).setOnClickListener(new OnClickListener() {
			
			@Override
			public void onClick(View v) {
				wimple.getFinancialState(DateFormatUtils.getServerDateString(""), true);
				llUpdateNotice.setVisibility(View.VISIBLE);
			}
		});
			
		return view;
	}
	@Override
	public void onDestroy() {

		/*
		asList.clear();
		asList = null;
		asAdapter.clear();
		asAdapter = null;
		 */

		llChart = null;
		tvSavingValue = null;
		tvDebtValue = null;
		tvSumValue = null;

		super.onDestroy();
	}
	@Override
	public void onDetach() {

		super.onDetach();
	}
	@Override
	public void onResume() {
		context = WimpleActivity.context.get();
		wimple.getFinancialState(DateFormatUtils.getServerDateString(""), false);
		super.onResume();
	}
	@SuppressWarnings({"unchecked", "deprecation"})
	@Override
	public void handleMessage(Message msg) {
		int command = msg.what;
		boolean booleanStatus = msg.arg1 == 1;
		Object obj = msg.obj;

		// if fragment is added or not to the activity
		if(!isAdded()){
			return;
		}
		
		if(null == context){
			context = WimpleActivity.context.get();
			if(null == context){
				return;
			}
		}

		switch(command){

		case CommandID.GET_FINANCIAL_STATE_RESPONSE_RECEIVED :{

			llUpdateNotice.setVisibility(View.GONE);
			
			if(firstUpdate){
				firstUpdate = false;
				// To show previous data during new data dispatching without any GUI display delay.
				SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(context);
				boolean autoRefresh = sharedPref.getBoolean(SettingsFragment.KEY_FINANCIAL_STATE_AUTO_REFRESH, true);
				if(autoRefresh){
					wimple.getFinancialState(DateFormatUtils.getServerDateString(""), true);	
					llUpdateNotice.setVisibility(View.VISIBLE);
				}
			}

			if(!booleanStatus){
				return;
			}
			
			Double saving = 0.0;
			Double debt = 0.0;

			Collection<AccountState> accountStates = (Collection<AccountState>)obj;
			for(AccountState as : accountStates){

				if(!as.getGroup()){
					if(as.getCategory().startsWith("li")){
						debt += as.getAmount();
					}if(as.getCategory().startsWith("as")){
						saving += as.getAmount();
					}
				}/*else{
					//asAdapter.get().addAccountState(as);
				}*/
			}
			//asAdapter.get().notifyDataSetChanged();

			if(null == llChart){
				llChart = (LinearLayout) view.findViewById(R.id.chart);
			}

			tvSavingValue.setText(DateFormatUtils.getNoPointDecimalFormat().format(saving));
			tvDebtValue.setText(DateFormatUtils.getNoPointDecimalFormat().format(-debt));

			Double sum = saving - debt;

			tvSumValue.setText(DateFormatUtils.getNoPointDecimalFormat().format(sum));
			if(sum >= 0){
				tvSumValue.setTextColor(getResources().getColor(R.color.text_blue));
			}else{
				tvSumValue.setTextColor(getResources().getColor(R.color.text_red));
			}

			PieChart pcv = ChartUtils.makeChart(context,
					new double[] {
							Math.abs(saving),
							Math.abs(debt)
					},
					new String[] {
							getResources().getString(R.string.title_saving),
							getResources().getString(R.string.title_debt)
					},
					Math.abs(saving) > Math.abs(debt)? Math.abs(saving):Math.abs(debt));

			llChart.removeAllViews();
			llChart.addView(pcv, new LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.MATCH_PARENT));
		}
		break;
		}
	}

	@Override
	public void setActivityInstance(WimpleActivity instance) {
	}
}

File: wimple/app/src/main/java/kr/blogspot/charlie0301/wimple/FinancialStateSummaryFragment.kt
package kr.blogspot.charlie0301.wimple

import android.os.Bundle
import android.os.Message
import androidx.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams
import androidx.core.content.ContextCompat
import kr.blogspot.charlie0301.wimple.databinding.FragmentFinalcialStateSummaryTabBinding
import kr.blogspot.charlie0301.wimple.impl.WimpleImpl
import kr.blogspot.charlie0301.wimple.impl.util.ChartUtils
import kr.blogspot.charlie0301.wimple.impl.util.DateFormatUtils
import kr.blogspot.charlie0301.wimple.model.AccountState

class FinancialStateSummaryFragment : androidx.fragment.app.Fragment(), IWimpleFragment {

    private val wimple = WimpleImpl.getInstance()
    private var _binding: FragmentFinalcialStateSummaryTabBinding? = null
    private val binding get() = _binding!!

    // Data
    private var firstUpdate: Boolean = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _binding = FragmentFinalcialStateSummaryTabBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        this.firstUpdate = true
        this.wimple.getFinancialState(DateFormatUtils.getServerDateString(""), false)

        binding.asRefresh.setOnClickListener {
            this.wimple.getFinancialState(DateFormatUtils.getServerDateString(""), true)
            binding.asUpdateNotification.visibility = View.VISIBLE
        }
    }

    private fun setShowingNotification(show: Boolean) {
        _binding?.let {
            it.asUpdateNotification.visibility = if (show) View.VISIBLE else View.GONE
        }
    }

    override fun onResume() {
        this.wimple.getFinancialState(DateFormatUtils.getServerDateString(""), false)
        super.onResume()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun handleMessage(msg: Message) {
        val command = msg.what
        val booleanStatus = msg.arg1 == 1
        val obj = msg.obj

        // if fragment is added or not to the activity
        if (!this.isAdded || _binding == null) {
            return
        }

        if (null == this.context) {
            return
        }

        when (command) {

            CommandID.GET_FINANCIAL_STATE_RESPONSE_RECEIVED -> {

                setShowingNotification(false)

                if (this.firstUpdate) {
                    this.firstUpdate = false
                    this.wimple.getFinancialState(DateFormatUtils.getServerDateString(""), true)
                    setShowingNotification(true)
                }

                if (!booleanStatus) {
                    return
                }

                var saving = 0.0
                var debt = 0.0

                @Suppress("UNCHECKED_CAST") val accountStates = obj as Collection<AccountState>
                for (acs in accountStates) {

                    if (!acs.group) {
                        if (acs.category.startsWith("li")) {
                            debt += acs.amount
                        }
                        if (acs.category.startsWith("as")) {
                            saving += acs.amount
                        }
                    }/*else{
					//asAdapter.get().addAccountState(as);
				}*/
                }
                //asAdapter.get().notifyDataSetChanged();

                binding.asSavingValue.text = DateFormatUtils.getNoPointDecimalFormat().format(saving)
                binding.asDebtValue.text = DateFormatUtils.getNoPointDecimalFormat().format(-1 * debt)

                val sum = saving - debt

                binding.asSumValue.text = DateFormatUtils.getNoPointDecimalFormat().format(sum)
                if (sum >= 0) {
                    binding.asSumValue.setTextColor(ContextCompat.getColor(this.requireContext(), R.color.text_blue))
                } else {
                    binding.asSumValue.setTextColor(ContextCompat.getColor(this.requireContext(), R.color.text_red))
                }

                val pcv = ChartUtils.makeChart(this.context,
                        doubleArrayOf(Math.abs(saving), Math.abs(debt)),
                        arrayOf(this.resources.getString(R.string.title_saving), this.resources.getString(R.string.title_debt)),
                        if (Math.abs(saving) > Math.abs(debt)) Math.abs(saving) else Math.abs(debt))

                binding.chart.removeAllViews()
                binding.chart.addView(pcv, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
            }
        }
    }

    override fun setActivityInstance(instance: WimpleActivity) {}
}